src/test/ui/drop/issue-90752-raw-ptr-shenanigans.rs

```diff
@@ -0,0 +1,41 @@
+// run-pass
+
+use std::cell::RefCell;
+
+struct S<'a>(i32, &'a RefCell<Vec<i32>>);
+
+impl<'a> Drop for S<'a> {
+    fn drop(&mut self) {
+        self.1.borrow_mut().push(self.0);
+    }
+}
+
+fn test(drops: &RefCell<Vec<i32>>) {
+    let mut foo = None;
+    let pfoo: *mut _ = &mut foo;
+
+    match foo {
+        None => (),
+        _ => return,
+    }
+
+    // Both S(0) and S(1) should be dropped, but aren't.
+    unsafe { *pfoo = Some((S(0, drops), S(1, drops))); }
+
+    match foo {
+        Some((_x, _)) => {}
+        _ => {}
+    }
+}
+
+fn main() {
+    let drops = RefCell::new(Vec::new());
+    test(&drops);
+
+    // Ideally, we want this...
+    //assert_eq!(*drops.borrow(), &[0, 1]);
+
+    // But the delayed access through the raw pointer confuses drop elaboration,
```

Comment:
This is unfortunate. Do you have an idea about how this could be fixed in the future?

Reply:
The way we typically handle this is by assuming that any variable which has been mutably borrowed at a prior point in the CFG could change at any subsequent point. The correctness of that approach doesn't depend on stacked borrows.

We can't implement it directly on top of `MaybeInitializedPlaces`, however, since it would break code that moves out of a variable after taking a mutable reference to it. We would need two state vectors like #90214 has, but only in drop elaboration, where the increased precision from tracking variants is worthwhile.